plasma/position.go

```diff
@@ -0,0 +1,192 @@
+package plasma
+
+import (
+	"fmt"
+	"github.com/ethereum/go-ethereum/rlp"
+	"io"
+	"math/big"
+	"strconv"
+	"strings"
+)
+
+// Position for an input/output
+type Position struct {
+	BlockNum     *big.Int
+	TxIndex      uint16
+	OutputIndex  uint8
+	DepositNonce *big.Int
+}
+
+type position struct {
+	BlockNum     []byte
+	TxIndex      uint16
+	OutputIndex  uint8
+	DepositNonce []byte
+}
+
+const (
+	blockIndexFactor = 1000000
+	txIndexFactor    = 10
+)
+
+func NewPosition(blkNum *big.Int, txIndex uint16, oIndex uint8, depositNonce *big.Int) Position {
+	if depositNonce == nil {
+		depositNonce = big.NewInt(0)
+	}
+	if blkNum == nil {
+		blkNum = big.NewInt(0)
+	}
+
+	return Position{
+		BlockNum:     blkNum,
+		TxIndex:      txIndex,
+		OutputIndex:  oIndex,
+		DepositNonce: depositNonce,
+	}
+}
+
+func (p *Position) EncodeRLP(w io.Writer) error {
+	if p.BlockNum == nil {
+		p.BlockNum = big.NewInt(0)
+	}
+	if p.DepositNonce == nil {
+		p.DepositNonce = big.NewInt(0)
+	}
+	pos := position{p.BlockNum.Bytes(), p.TxIndex, p.OutputIndex, p.DepositNonce.Bytes()}
+
+	return rlp.Encode(w, pos)
+}
+
+func (p *Position) DecodeRLP(s *rlp.Stream) error {
+	var pos position
+	if err := s.Decode(&pos); err != nil {
+		return err
+	}
+
+	p.BlockNum = new(big.Int).SetBytes(pos.BlockNum)
+	p.TxIndex = pos.TxIndex
+	p.OutputIndex = pos.OutputIndex
+	p.DepositNonce = new(big.Int).SetBytes(pos.DepositNonce)
+
+	return nil
+}
+
+func (p Position) Bytes() []byte {
+	bytes, _ := rlp.EncodeToBytes(&p)
+	return bytes
+}
+
+// ValidateBasic ensures deposit and child chain positions are mutually exclusive.
+// Block numbering starts at 1
+func (p Position) ValidateBasic() error {
+	if p.IsNilPosition() {
+		return fmt.Errorf("nil position is not a valid position")
+	}
+
+	// deposit position
+	if p.IsDeposit() {
+		if p.BlockNum.Sign() > 0 || p.TxIndex > 0 || p.OutputIndex > 0 {
+			return fmt.Errorf("chain postion must be all zero if a deposit nonce is specified. (0.0.0.nonce)")
+		}
+	} else {
+		if p.BlockNum.Sign() == 0 {
```

Comment:
I believe we need a check to make sure TxIndex < 2 ** 16 and OIndex < 2 to prevent priority collisions